test/common/router/retry_state_impl_test.cc

```diff
@@ -704,16 +704,32 @@ TEST_F(RouterRetryStateImplTest, PolicyLimitedByRequestHeaders) {
     EXPECT_FALSE(state_->enabled());
   }
 
+  {
+    Http::TestHeaderMapImpl request_headers{{":method", "GET"},
+                                            {"x-envoy-retry-on", "retriable-4xx"}};
+    setup(request_headers);
+    EXPECT_TRUE(state_->enabled());
+    Http::TestHeaderMapImpl response_headers{{":status", "409"}};
+    expectTimerCreateAndEnable();
+    EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
+  }
+
   {
     Http::TestHeaderMapImpl request_headers{{":method", "GET"}, {"x-envoy-retry-on", "5xx"}};
     setup(request_headers);
     EXPECT_TRUE(state_->enabled());
+    Http::TestHeaderMapImpl response_headers{{":status", "500"}};
+    expectTimerCreateAndEnable();
+    EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
   }
 
   {
     Http::TestHeaderMapImpl request_headers{{":method", "HEAD"}, {"x-envoy-retry-on", "5xx"}};
     setup(request_headers);
     EXPECT_TRUE(state_->enabled());
+    Http::TestHeaderMapImpl response_headers{{":status", "500"}};
+    expectTimerCreateAndEnable();
+    EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
```

Comment:
Should we also have a test where the policy is enabled but the RetryStatus is No?

Reply:
sure ill add one